clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3950,13 +3950,23 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
   }
 
   // Default inits the type, then calls the init-method in the body.
+  // A type may not have a public default constructor as per its spec so
+  // typically if this is the case the default constructor will be private and
+  // in such cases we must manually override the access specifier from private
+  // to public just for the duration of this default initialization.
   bool handleSpecialType(FieldDecl *FD, QualType Ty) {
+    const auto *RecordDecl = Ty->getAsCXXRecordDecl();
+    AccessSpecifier DefaultConstructorAccess;
+    auto DefaultConstructor =
+        std::find_if(RecordDecl->ctor_begin(), RecordDecl->ctor_end(),
+                     [](auto it) { return it->isDefaultConstructor(); });
+    DefaultConstructorAccess = DefaultConstructor->getAccess();
+    DefaultConstructor->setAccess(AS_public);
     addFieldInit(FD, Ty, std::nullopt,
                  InitializationKind::CreateDefault(KernelCallerSrcLoc));
-
+    DefaultConstructor->setAccess(DefaultConstructorAccess);
```

Comment:
Does it actually complain during kernel construction if you don't set access?
If yes, does it complain when it should after this change?

Reply:
> Does it actually complain during kernel construction if you don't set access? If yes, does it complain when it should after this change?

Yes, during the construction of the SYCL Kernel Body, it complains that the constructor has an incorrect number of arguments because it doesn't see a public default constructor. After this change, it doesn't complain anymore. Its also an approach for any special type so we don't have to change our handling based on the special type.

Reply:
> Does it actually complain during kernel construction if you don't set access? If yes, does it complain when it should after this change?

This line in particular that you have reviewed sets the access back to its original value so that we don't accidentally leave the default constructor visible when it shouldn't be. Frankly, even without this line it seemed like the user could still not call the default constructor if it was defined private, but I chose to err on the safe side.

Reply:
> I see. I suppose I would also prefer the safe side. Shouldn't you be doing the same for each variant of `handleSpecialType`?

Good question. This change is only relevant for the semantic action of the SyclKernelBodyCreator class which has 3 members named `handleSpecialType`. One is for kernel handler which has its own thing going on so I don't think it needs these changes, the other is the one I've changed here and the remaining one is the one taking a `CXXBaseSpecifier`. I suppose I should add it to this one too but I forgot, is that what you were asking about?

Reply:
> I suppose I should add it to this one too but I forgot, is that what you were asking about?

Right. Passing precommit also suggests lack of corresponding tests. I suppose a test should contain work group memory as a base to exercise the code.

Reply:
Ok. I will add such a test.

Reply:
I do not have another solution but this just feels very hacky and I really doubt community clang will accept this. This is out of the scope of the PR but the design of how we handle initialization of special types probably needs to change if we cannot use a default constructor. @tahonermann please weigh in here.